### DISCUSSION TITLE
Typescript discriminating data based on isLoading state

### DISCUSSION THREAD

**Discussion by tomchambers2** [reduxjs/redux-toolkit#1475] [Ideas]
Currently you need to keep checking if data exists in order to use it. Would it be possible to set the types so that we can use the `data` property safely if `isLoading` or `isFetching` has been checked?

A type that looks like
```
type HookResult = {
    data: undefined;
    isLoading: true;
} | {
    data: Data;
    isLoading: false;
}
```

So you could do
```
const { data, isLoading } = useGetSomethingQuery(args);

if (isLoading) return <div>Loading...</div>;

<div>{data}</div>
```

(From memory) I think that react-query does something like this?

**Comment by phryneas** [comment]
It is set up that way. But when you destructure, the TypeScript compiler loses that information. Try to keep it as one object instead of destructuring.

**Reply by tomchambers2** [reply]
Thanks for the explanation
